python_modules/dagster/dagster/_grpc/server.py

```diff
@@ -113,11 +117,49 @@
 
 STREAMING_CHUNK_SIZE = 4000000
 
+_UTILIZATION_METRICS = defaultdict(dict)
+_METRICS_LOCK = threading.Lock()
+METRICS_RETRIEVAL_FUNCTIONS = set()
+
+
+def _gen_request_id():
+    return str(ulid.new().uuid)
+
 
 class CouldNotBindGrpcServerToAddress(Exception):
     pass
 
 
+def retrieve_metrics(arg_from_request: str):
+    class _MetricsRetriever:
```

Comment:
This follows a similar pattern to what we use for our other decorators, IE `op`, `asset`, etc. which allows us to provide arguments to the decorator, while still using it in such a manner.